tools/SourceKit/lib/SwiftLang/SwiftASTManager.h

```diff
@@ -68,24 +132,97 @@ class ASTUnit : public SourceKit::ThreadSafeRefCountedBase<ASTUnit> {
 
 typedef IntrusiveRefCntPtr<ASTUnit> ASTUnitRef;
 
-class SwiftASTConsumer {
+class SwiftASTConsumer : public std::enable_shared_from_this<SwiftASTConsumer> {
+  /// Mutex guarding all accesses to CancellationRequestCallback
+  llvm::sys::Mutex CancellationRequestCallbackMtx;
+
+  /// A callback that informs the \c ASTBuildOperation, which is producing the
+  /// AST for this consumer, that the consumer is no longer of interest. Calling
+  /// this callback will eventually call \c cancelled on this consumer.
+  /// If the consumer isn't associated with any \c ASTBuildOperation at the
+  /// moment (e.g. if it hasn't been scheduled on one yet or if the build
+  /// operation has already informed the ASTConsumer), the callback is \c None.
+  Optional<std::function<void(std::shared_ptr<SwiftASTConsumer>)>>
+      CancellationRequestCallback;
+
 public:
   virtual ~SwiftASTConsumer() { }
+
+  // MARK: Cancellation
+
+  /// The result of this consumer is no longer of interest to the SourceKit
+  /// client.
+  /// This will cause \c cancelled to be called on this \c SwiftASTConsumer and
+  /// cause the \c ASTBuildOperation to be cancelled if no other consumer is
+  /// depending on it.
+  void requestCancellation() {
+    CancellationRequestCallbackMtx.lock();
+    if (CancellationRequestCallback.hasValue()) {
+      // Copy the function to a new variable so we can call it thread-safe
+      // without CancellationRequestCallbackMtx being locked. This allows the
+      // callback to remove itself from this SwiftASTConsumer.
+      std::function<void(std::shared_ptr<SwiftASTConsumer>)>
+          CancellationRequest = *CancellationRequestCallback;
+      CancellationRequestCallbackMtx.unlock();
+      CancellationRequest(shared_from_this());
+    } else {
+      CancellationRequestCallbackMtx.unlock();
+    }
+  }
+
+  /// Set a cancellation request callback that informs a \c ASTBuildOperation
+  /// when this \c SwiftASTConsumer is cancelled. Asserts that there is
+  /// currently no callback set.
+  void setCancellationRequestCallback(
+      std::function<void(std::shared_ptr<SwiftASTConsumer>)> NewCallback) {
+    llvm::sys::ScopedLock L(CancellationRequestCallbackMtx);
+    assert(!CancellationRequestCallback.hasValue() &&
+           "Can't set two cancellation callbacks on a SwiftASTConsumer");
+    CancellationRequestCallback = NewCallback;
+  }
+
+  /// Removes the cancellation request callback previously set by \c
+  /// setCancellationRequestCallback.
+  void removeCancellationRequestCallback() {
+    llvm::sys::ScopedLock L(CancellationRequestCallbackMtx);
+    CancellationRequestCallback = None;
+  }
+
+  /// MARK: Result methods
+  /// Exactly one of these will be called for every consumer.
```

Comment:
This should not be a doc comment since it's not attached to anything.  Also in clang the spelling is `#pragma mark` instead of `// MARK:`.

Reply:
Yeah, this shouldn’t be a doc comment. `// MARK: ` is also recognized by Xcode for C++ and we are using it in quite a few places already. Since I like it better than the `#pragma mark`, I’d prefer to stick with `// MARK:`.